### PR TITLE
Support splat expansions inside tuple and array literals

### PR DESCRIPTION
**Pull request by HertzDevil**
Implements #3718. Does not implement double splats and https://github.com/crystal-lang/crystal/issues/3718#issuecomment-782248824 (`#to_array_splat`). The codegen portion for `Tuple` splats is copied directly from #10193's [head](https://github.com/crystal-lang/crystal/pull/10193/commits/9ed19b90976e03a25728a8bc7dc3d4bd871791f4) (with the codegen optimization).

Tuple and array literals have slightly different requirements on the splatted value's interface:

```crystal
# `exp` must be a `Tuple` instance, it cannot be a union of `Tuple`s (of different arities)
{*exp}

# `exp` must respond to `#each(& : _ -> _)`
[*exp]
[*exp] of T
T{*exp}
T(U, ...){*exp}
```

~~Very few array-like types implement `#each` but do not include `Enumerable`, so the requirement on `Enumerable` shouldn't be a pressing issue. (`CSV` and `Regex::MatchData` are notable exceptions in the standard library.)~~

### CODE REVIEW

File: src/compiler/crystal/syntax/parser.cr [hunk 1]
```diff
@@ -2320,7 +2320,16 @@ module Crystal
...
-          exps << parse_op_assign_no_control
+          exp_location = @token.location
+          has_splat = false
+          if @token.type == :"*"
+            has_splat = true
+            next_token_skip_space_or_newline
+          end
+
+          exp = parse_op_assign_no_control
+          exp = Splat.new(exp).at(exp_location) if has_splat
```
Comment (straight-shoota) [inline src/compiler/crystal/syntax/parser.cr:2331]: Minor structural suggestion, but I think this code is better with two separate branches. The only shared expression is a single call.

```suggestion
          if @token.type == :"*"
            next_token_skip_space_or_newline
            location = @token.location
            exp = Splat.new(parse_op_assign_no_control).at(location)
          else
            exp = parse_op_assign_no_control
          end
```
Ditto below.

**Review by straight-shoota — CHANGES_REQUESTED** [review]

**Comment by HertzDevil** [comment]
Instead of `__crystal_first` there is now `Enumerable.element_type`, other kinds of literal expansions refer to class methods in the prelude so this shouldn't be a problem. It might replace the `first_internal` helper I added in some other PRs too.

**Review by straight-shoota — APPROVED** [review]

File: src/enumerable.cr [hunk 2]
```diff
@@ -1911,4 +1911,21 @@ module Enumerable(T)
...
+  #
+  # Used by splat expansion inside array literals. For example, this code
+  #
+  # ```
+  # [1, *{2, 3.5}, 4]
+  # ```
+  #
+  # will end up calling `typeof(1, ::Enumerable.element_type({2, 3.5}), 4)`.
+  #
+  # NOTE: there should never be a need to call this method outside the standard library.
```
[the remark this answers is not in the document]
Comment (straight-shoota) [reply src/enumerable.cr:1926]: I don't think this should be a warning.
We could consider making this method nodoc, though. It's only internal API.

**Comment by HertzDevil** [comment]
The PR currently doesn't allow non-`Enumerable` splats inside `Array`s because of `Array#concat(other : Enumerable)`'s type restriction (array literals expand to `concat`, not the more verbose `each` loop). Should we remove that type restriction?

**Comment by straight-shoota** [comment]
I suppose this restriction is not really necessary. However, what type would implement a compatible `#each` interface an *not* include `Enumerable`? Practically, I don't see much impact.

**Comment by HertzDevil** [comment]
`CSV`, `Enum` (flags enums only), `LLVM::FunctionCollection`, and `YAML::Nodes::Mapping`.

File: src/compiler/crystal/codegen/codegen.cr [hunk 3]
```diff
@@ -476,10 +476,35 @@ module Crystal
     def visit(node : TupleLiteral)
       request_value do
         type = node.type.as(TupleInstanceType)
-        @last = allocate_tuple(type) do |tuple_type, i|
-          exp = node.elements[i]
-          accept exp
-          {exp.type, @last}
+
+        if node.elements.any?(Splat)
```
Comment (asterite) [inline src/compiler/crystal/codegen/codegen.cr:480]: Great use of this variant of `any?`! It actually reads really nice too :-)

**Review by asterite — APPROVED** [review]
I love this new feature! ❤️

**Pull request merged**